ipam/monitor/awsvpc.go

```diff
@@ -0,0 +1,265 @@
+	"github.com/weaveworks/weave/common"
+	"github.com/weaveworks/weave/net/address"
+)
+
+type AWSVPCMonitor struct {
+	ec2          *ec2.EC2
+	instanceID   string
+	routeTableID string
+	linkIndex    int
+}
+
+const (
+	bridgeIfName = "weave"
+)
+
+// NewAWSVPCMonitor creates and initialises AWS VPC based monitor.
+//
+// The monitor updates AWS VPC and host route tables when any changes to allocated
+// address ranges owner by a peer have been committed.
+func NewAWSVPCMonitor() (*AWSVPCMonitor, error) {
+	var err error
+	session := session.New()
+	mon := &AWSVPCMonitor{}
+
+	// Detect region and instance id
+	meta := ec2metadata.New(session)
+	mon.instanceID, err = meta.GetMetadata("instance-id")
+	if err != nil {
+		return nil, fmt.Errorf("cannot detect instance-id: %s", err)
+	}
+	region, err := meta.Region()
+	if err != nil {
+		return nil, fmt.Errorf("cannot detect region: %s", err)
+	}
+
+	mon.ec2 = ec2.New(session, aws.NewConfig().WithRegion(region))
+
+	routeTableID, err := mon.detectRouteTableID()
+	if err != nil {
+		return nil, err
+	}
+	mon.routeTableID = *routeTableID
+
+	// Detect Weave bridge link index
+	link, err := netlink.LinkByName(bridgeIfName)
+	if err != nil {
+		return nil, fmt.Errorf("cannot find \"%s\" interface: %s", bridgeIfName, err)
+	}
+	mon.linkIndex = link.Attrs().Index
+
+	common.Log.Debugf(
+		"AWSVPC monitor has been initialized on %s instance for %s route table at %s region",
+		mon.instanceID, mon.routeTableID, region)
+
+	return mon, nil
+}
+
+// HandleUpdate method updates the AWS VPC and the host route tables.
+func (mon *AWSVPCMonitor) HandleUpdate(old, new []address.Range) error {
+	oldCIDRs, newCIDRs := filterOutSameCIDRs(address.NewCIDRs(old), address.NewCIDRs(new))
+	common.Log.Debugf("HandleUpdate: old(%q) new(%q)", old, new)
+
+	// It might make sense to do removal first and then add entries
+	// because of the 50 routes limit. However, in such case a container might
+	// not be reachable for short period of time which we we would like to
+	// avoid.
+
+	// Add new entries
+	for _, cidr := range newCIDRs {
+		cidrStr := cidr.String()
+		common.Log.Debugf("Creating %s route to %s", cidrStr, mon.instanceID)
+		_, err := mon.createVPCRoute(cidrStr)
+		// TODO(mp) check for 50 routes limit
+		// TODO(mp) maybe check for auth related errors
+		if err != nil {
+			return fmt.Errorf("createVPCRoutes failed: %s", err)
+		}
+		err = mon.createHostRoute(cidrStr)
+		if err != nil {
+			return fmt.Errorf("createHostRoute failed: %s", err)
+		}
+	}
+
+	// Remove obsolete entries
+	for _, cidr := range oldCIDRs {
+		cidrStr := cidr.String()
+		common.Log.Debugf("Removing %s route", cidrStr)
+		_, err := mon.deleteVPCRoute(cidrStr)
+		if err != nil {
+			return fmt.Errorf("deleteVPCRoute failed: %s", err)
+		}
+		err = mon.deleteHostRoute(cidrStr)
+		if err != nil {
+			return fmt.Errorf("deleteHostRoute failed: %s", err)
+		}
+	}
+
+	return nil
+}
+
+func (mon *AWSVPCMonitor) String() string {
+	return "awsvpc"
```

Comment:
Used here: https://github.com/weaveworks/weave/blob/issues/2017-awsvpc-wip/ipam/status.go#L17 which is used to detect whether we use VPC when attaching containers in https://github.com/weaveworks/weave/blob/issues/2017-awsvpc-wip/weave#L1014